api/swagger.yaml

```diff
@@ -1363,16 +1363,16 @@ definitions:
       EndpointsConfig:
         description: |
           A mapping of network name to endpoint configuration for that network.
+          The endpoint configuration can be left empty to connect to that
+          network with no particular endpoint configuration.
         type: "object"
         additionalProperties:
           $ref: "#/definitions/EndpointSettings"
     example:
       # putting an example here, instead of using the example values from
-      # /definitions/EndpointSettings, because containers/create currently
-      # does not support attaching to multiple networks, so the example request
-      # would be confusing if it showed that multiple networks can be contained
-      # in the EndpointsConfig.
-      # TODO remove once we support multiple networks on container create (see https://github.com/moby/moby/blob/07e6b843594e061f82baa5fa23c2ff7d536c2a05/daemon/create.go#L323)
+      # /definitions/EndpointSettings, because EndpointSettings contains
+      # operational data returned when inspecting a container that we don't
+      # accept here.
```

Comment:
Do you know what data we have in `"#/definitions/EndpointSettings"` that doesn't apply here?

Funny thing is that I don't think this example is actually used, because the container-create endpoint defines a "full" example for creating containers (including NetworkSettings); https://github.com/moby/moby/blob/462d6ef826861fad021fb565c0481fb61d2db6bc/api/swagger.yaml#L6409-L6423

Reply:
> Do you know what data we have in "#/definitions/EndpointSettings" that doesn't apply here?

Assuming this example was really used for documenting `docker create` and `docker network connect` (which is not the case, as you pointed out):

https://github.com/moby/moby/blob/462d6ef826861fad021fb565c0481fb61d2db6bc/api/swagger.yaml#L2464-L2521

I started looking at whether we could move these operational data from `EndpointSettings` into a dedicated `EndpointState`, but this is going to be high-effort/low-impact so I'd prefer to focus on more pressing issues first.

Reply:
Yeah, there's still a lot of cleaning up to do.

I've been slowly chipping away some of that; my attempts there are mostly to use _per field_ `example` values. Reason for that is that the swagger docs present example responses as-is, which also means that they don't have to match the actual type returned ... at all (!).

That has lead to examples not matching reality on various occasions, e.g. the following just happily renders a completely bogus response;

```yaml
swagger: "2.0"
info:
  title: "Foobar API"
  version: "1.0.0"

definitions:
  MyType:
    type: "object"
    properties:
      Hello:
        type: "string"
        example: "hello"
      World:
        type: "string"
        example: "world"

paths:
  /containers/json:
    get:
      responses:
        200:
          description: "no error"
          schema:
            $ref: "#/definitions/MyType"
          examples:
            application/json:
              - This: "this"
                Does: "does"
                Not:
                  Have: "have"
                  To: "match"
```

<img width="1264" alt="Screenshot 2023-07-11 at 13 39 25" src="https://github.com/moby/moby/assets/1804568/2e9c5ca1-1c01-4501-aa47-470f471a0360">


The _downside_ on the other hand, is that all fields in the definition need to have some sample value (doable), but due to how we use the same structs for different purposes, and some fields being "optional", the examples could sometimes contain fields set that should not be set (e.g. container _create_ would not contain an ID, whereas container _inspect_ would contain it (as it was generated by the daemon).

So we'd have to split definitions for those, or find some other way to reflect that.

Reply:
> some fields being "optional", the examples could sometimes contain fields set that should not be set (e.g. container create would not contain an ID, whereas container inspect would contain it (as it was generated by the daemon).

`EndpointSettings` is actually one of these. All these fields here should not be sent on `ContainerCreate`:

https://github.com/moby/moby/blob/b2bde4a7d8696943a51e1eca44735841d5c159e1/api/types/network/network.go#L54-L64

As suggested in this comment https://github.com/moby/moby/pull/46059#discussion_r1272152754, I would really like to see this struct split in two: one struct for the user-provided settings, and the other tracking the endpoint state.